global.go

```diff
@@ -10,17 +10,54 @@ import (
+//
+// At the beginning of the optimization, InitGlobal is called to communicate
+// the dimension of the input and maximum number of concurrent tasks.
+// The actual number of concurrent tasks will be set from the return of InitGlobal,
+// which must not be greater than the input tasks.
+//
+// During the optimization, a reverse-communication interface is used between
+// the GlobalMethod and the caller.
+// GlobalMethod acts as a client that asks the caller to perform
+// needed operations given the return from IterateGlobal.
+// This allows and enforces automation of maintaining statistics and checking for
+// (various types of) convergence.
+//
+// The return from IterateGlobal can be an Evaluation, a MajorIteration or NoOperation.
+//
+// An evaluation is one or more of the Evaluation operations (FuncEvaluation,
+// GradEvaluation, etc.) combined with the bitwise or operator. In an evaluation
+// operation, the requested fields of Problem will be evaluated at the value
+// in Location.X, filling the corresponding fields of Location. These values
+// can be retrieved and used upon the next call to IterateGlobal with that task id.
+// The corresponding fields of Location will be filled with the results that
+// The GlobalMethod interface requires that entries of Location are not modified
+// aside from the commanded evaluations. Thus, the type implementing Method may
+// use multiple Operations to set the Location fields at a particular x value.
+//
+// A GlobalMethod may alternately declare MajorIteration. This updates the optimal
+// location to the values in Location, and checks for the convergence of the
+// optimization. Consequently, the fields of Location must be valid and consistent.
+//
+// A GlobalMethod must not return InitIteration and PostIteration operations. These are
+// reserved for the clients to be passed to Recorders. A Method must also not
+// combine the Evaluation operations with the Iteration operations.
 type GlobalMethod interface {
-	// Global tells method the max number of tasks, method returns how many it wants.
-	// This is needed to sync the Global goroutines and inside goroutines.
+	Needser
+	// InitGlobal communicates the input dimension and maximum number of tasks,
+	// and returns the number of concurrent processes. The return must be less
+	// than or equal to tasks.
 	InitGlobal(dim, tasks int) int
-	// Global method may assume that the same task id always has the same pointer with it.
+
+	// IterateGlobal retrieves information from the location associated with
+	// the given task ID, and returns the next operation to perform with that
+	// Location. IterateGlobal may assume that the same pointer is associated
+	// with the same task.
 	IterateGlobal(task int, loc *Location) (Operation, error)
-	Needser
-	// Done communicates to the optimization method that the optimization has
-	// concluded to allow for shutdown.
+
+	// Done communicates that the optimization has concluded to allow for shutdown.
```

Comment:
perhaps semantically relate it to the `sync.WaitGroup.Done()` method? (or explicitly unrelate it to that)

I was a bit thrown off tracks that it didn't return anything (not even `struct{}`) so it couldn't really be used like in:
```go
for {
    select {
    case <-gm.Done(): // do clean-up
    }
}
```
and then I remembered that it looked a bit like `sync.WaitGroup.Done()`.